middle_end/flambda/from_lambda/cps_conversion.ml

```diff
@@ -27,40 +27,251 @@ module C = Lambda_conversions
+      static_exn_continuation = Numbers.Int.Map.empty;
+      recursive_static_catches = Numbers.Int.Set.empty;
     }
 
   let current_unit_id t = t.current_unit_id
 
-  let add_mutable t id =
-    assert (not (Ident.Set.mem id t.mutables));
-    t.has_mutables := true;
-    { t with mutables = Ident.Set.add id t.mutables; }
-
   let is_mutable t id =
-    Ident.Set.mem id t.mutables
+    Ident.Map.mem id t.current_values_of_mutables_in_scope
 
   let has_mutables t =
     !(t.has_mutables)
+
+  let add_mutable_and_make_new_id t id kind =
+    if Ident.Map.mem id t.current_values_of_mutables_in_scope then begin
+      Misc.fatal_errorf "Redefinition of mutable variable %a"
+        Ident.print id
+    end;
+    let new_id = Ident.rename id in
+    let current_values_of_mutables_in_scope =
+      Ident.Map.add id (new_id, kind) t.current_values_of_mutables_in_scope
+    in
+    let t =
+      { t with
+        current_values_of_mutables_in_scope;
+      }
+    in
+    t, new_id
+
+  let new_id_for_mutable t id =
+    match Ident.Map.find id t.current_values_of_mutables_in_scope with
+    | exception Not_found ->
+      Misc.fatal_errorf "Mutable variable %a not in environment"
+        Ident.print id
+    | _old_id, kind ->
+      let new_id = Ident.rename id in
+      let current_values_of_mutables_in_scope =
+        Ident.Map.add id (new_id, kind) t.current_values_of_mutables_in_scope
+      in
+      let t =
+        { t with
+          current_values_of_mutables_in_scope;
+        }
+      in
+      t, new_id, kind
+
+  let mutables_in_scope t =
+    Ident.Map.keys t.current_values_of_mutables_in_scope
+
+  type add_continuation_result = {
+    body_env : t;
+    handler_env : t;
+    extra_params : (Ident.t * Lambda.value_kind) list;
+  }
+
+  let add_continuation t cont ~push_to_try_stack
+    (recursive : Asttypes.rec_flag) =
+    let body_env =
+      let mutables_needed_by_continuations =
+        Continuation.Map.add cont (mutables_in_scope t)
+          t.mutables_needed_by_continuations
+      in
+      let try_stack =
+        if push_to_try_stack then cont::t.try_stack else t.try_stack
+      in
+      { t with
+        mutables_needed_by_continuations;
+        try_stack;
+      }
+    in
+    let current_values_of_mutables_in_scope =
+      Ident.Map.mapi (fun mut_var (_outer_value, kind) ->
+          Ident.rename mut_var, kind)
+        t.current_values_of_mutables_in_scope
+    in
+    let handler_env =
+      let handler_env =
+        match recursive with
+        | Nonrecursive -> t
+        | Recursive ->
+          if push_to_try_stack then begin
+            Misc.fatal_error "Try continuations should not be recursive"
+          end;
+          body_env
+      in
+      { handler_env with
+        current_values_of_mutables_in_scope;
+      }
+    in
+    let extra_params =
+      List.map snd
+        (Ident.Map.bindings handler_env.current_values_of_mutables_in_scope)
```

Comment:
You should be able to use `Ident.Map.data` directly (most maps in the compiler implement the signature in `identifiable.mli`, which has a few extra convenience functions).